util/entropy.c

```diff
@@ -16,61 +18,157 @@
  * XXX win32 where there is no concept of /dev/urandom.
  */
 
+struct entropy_read
+{
+	int fd;
+};
+
 /**
- * entropy_read(buf, buflen):
- * Fill the given buffer with random bytes provided by the operating system.
+ * entropy_read_open():
+ * Initialize the ability to produce random bytes, and return a cookie.
  */
-int
-entropy_read(uint8_t * buf, size_t buflen)
+static struct entropy_read *
+entropy_read_open(void)
 {
-	int fd;
-	ssize_t lenread;
+	static struct entropy_read * er;
 
-	/* Sanity-check the buffer size. */
-	if (buflen > SSIZE_MAX) {
-		warn0("Programmer error: "
-		    "Trying to read insane amount of random data: %zu",
-		    buflen);
+	/* Allocate cookie. */
+	if ((er = malloc(sizeof(struct entropy_read))) == NULL) {
+		warnp("malloc");
 		goto err0;
 	}
 
 	/* Open /dev/urandom. */
-	if ((fd = open("/dev/urandom", O_RDONLY)) == -1) {
+	if ((er->fd = open("/dev/urandom", O_RDONLY)) == -1) {
 		warnp("open(/dev/urandom)");
+		goto err1;
+	}
+
+	/* Success! */
+	return (er);
+
+err1:
+	free(er);
+err0:
+	/* Failure! */
+	return (NULL);
+}
+
+/**
+ * entropy_read_fill(er, buf, buflen):
+ * Fill the given buffer with random bytes provided by the operating system
+ * using the file descriptor in ${er}.
+ */
+static int
+entropy_read_fill(struct entropy_read * er, uint8_t * buf, size_t buflen)
+{
+	ssize_t lenread;
+
+	/* Sanity checks. */
+	assert(er != NULL);
+	if (buflen > SSIZE_MAX) {
```

Comment:
What would you think about replacing this with
```
assert(buflen <= SSIZE_MAX);
```

with the general idea of "use `assert()` for programming errors and don't bother trying to quit gracefully".

I note that `entropy.c` is the only file `.c` in libcperciva, scrypt, spiped, or kivaloo that uses the term "Programmer error".  (Admittedly there's a ton of them in tarsnap.)

Reply:
Switching to `assert` is good.  Some or all of the ones in tarsnap should probably be changed too.